Common/Exceptions/PythonUserExceptionParser.cs

```diff
@@ -13,58 +13,91 @@
+    /// Parser that converts a <see cref="PythonException"/> thrown by a python algorithm into an <see cref="Exception"/>
     /// </summary>
     public class PythonUserExceptionParser : IExceptionParser
     {
-        private static readonly Dictionary<string, string> _commonErrors = new Dictionary<string, string>
-        {
-            { "KeyError", "Trying to retrieve an element from a collection using a key that does not exist in that collection throws a KeyError exception. To prevent the exception, ensure that the key exist in the collection and/or that collection is not empty."},
-            { "UnsupportedOperandError", "Trying to perform a summation, subtraction, multiplication or division between a decimal.Decimal and a float throws a TypeError exception. To prevent the exception, ensure that both values share the same type, either decimal.Decimal or float."},
-            { "ZeroDivisionError", "Trying to divide an integer or Decimal number by zero throws a DivideByZeroException exception. To prevent the exception, ensure that the denominator in a division operation with integer or Decimal values is non-zero." },
-        };
-
         /// <summary>
-        /// Parses an <see cref="PythonException"/> object into an <see cref="UserException"/> one
+        /// Parses an <see cref="Exception"/> object into an new <see cref="Exception"/> with a legible message.
         /// </summary>
-        /// <param name="exception"><see cref="PythonException"/> object to parse into an <see cref="UserException"/> one.</param>
+        /// <param name="exception"><see cref="Exception"/> object to parse.</param>
         /// <returns>Parsed exception</returns>
         public Exception Parse(Exception exception)
         {
-            var pythonException = exception as PythonException;
-            if (pythonException == null)
+            var original = exception;
+            if (exception.InnerException != null)
             {
-                pythonException = exception.InnerException as PythonException;
-                if (pythonException == null)
-                {
-                    throw new ArgumentException("The given exception is not valid since it is of type PythonException nor its InnerException is");
-                }
+                exception = Parse(exception.InnerException);
+            }
+
+            var message = CreateLegibleMessage(exception.Message);
+            var errorLine = CreateErrorLine(exception.StackTrace);
+            message = $"{message}{errorLine}";
+
+            if (original.GetType() == typeof(InitializeException))
+            {
+                message = $"In the Initialize method, {message.Substring(0, 1).ToLower()}{message.Substring(1)}";
+                return new InitializeException(message, original);
+            }
+            else if (original.GetType() == typeof(ScheduledEventException))
```

Comment:
shame this is duplicated from the csharp type, let's not touch this now, but a note for our future selves that this can be improved.